File: sidecars/matter-server/src/protocol.ts
// Wire protocol between the Go core (keystone) and this sidecar.
// Must stay byte-compatible with internal/adapters/matter/protocol.go.

export interface RpcRequest {
    id: string;
    method: string;
    params?: unknown;
}

export interface RpcSuccess {
    id: string;
    result: unknown;
}

export interface RpcFailure {
    id: string;
    error: RpcErrorBody;
}

export interface ServerEvent {
    event: string;
    data: unknown;
    /**
     * Monotonic per-process sequence number, starting at 1. Lets a client
     * detect loss and ask for a replay after a reconnect.
     *
     * The counter resets when the sidecar restarts, so a client must treat a
     * `seq` lower than what it has already seen as "the server restarted",
     * not as a duplicate — see SubscribeResult.seq.
     */
    seq: number;
}

export type RpcResponse = RpcSuccess | RpcFailure;
export type OutboundFrame = RpcResponse | ServerEvent;

// --- method names ---
export const Methods = {
    Commission: "commission",
    ListNodes: "listNodes",
    ReadAttribute: "readAttribute",
    WriteAttribute: "writeAttribute",
    InvokeCommand: "invokeCommand",
    RemoveNode: "removeNode",
    Subscribe: "subscribe",
    DiscoverCommissionable: "discoverCommissionable",
    WebrtcOffer: "webrtcOffer",
    WebrtcIce: "webrtcIce",
    WebrtcStop: "webrtcStop",
} as const;

// --- event names ---
export const Events = {
    AttributeChanged: "attributeChanged",
    NodeOnline: "nodeOnline",
    NodeOffline: "nodeOffline",
    CommissioningProgress: "commissioningProgress",
    CommissionableFound: "commissionableFound",
    DeviceEvent: "deviceEvent",
    WebrtcSignal: "webrtcSignal",
} as const;

// --- method payloads ---

export interface CommissionParams {
    setupCode: string;
    /**
     * Ref of a device previously reported by discoverCommissionable. When set,
     * commissioning targets exactly that device and only the passcode is taken
     * from `setupCode`.
     *
     * Discovery never removes the need for a setup code: the passcode is not
     * advertised, so PASE cannot start without it. Targeting only saves the
     * user from disambiguating identical devices.
     */
    target?: string;
}

export interface DiscoverCommissionableParams {
    /** How long to listen for advertisements. Defaults to 10s, capped at 60s. */
    timeoutMs?: number;
}

/**
 * A device advertising itself as ready to be commissioned. Everything here
 * comes from the mDNS/BLE advertisement — the device has not been interviewed,
 * so there are no clusters or features yet.
 */
export interface CommissionableDevice {
    /** Stable handle for this advertisement; pass back as CommissionParams.target. */
    ref: string;
    /** Advertised device name, if the vendor set one. */
    name?: string;
    /** Resolved CSA device-type name ("OnOffLight"), when advertised and known. */
    deviceType?: string;
    vendorId?: number;
    productId?: number;
    discriminator?: number;
}

export interface CommissionResult {
    nodeId: string;
    fabricIndex: number;
}

export interface Node {
    nodeId: string;
    vendorName?: string;
    productName?: string;
    vendorId?: number;
    productId?: number;
    /** User-assigned name from the ecosystem that commissioned the device. */
    nodeLabel?: string;
    online: boolean;
    endpoints: Endpoint[];
}

export interface Endpoint {
    endpointId: number;
    deviceType?: string;
    clusters: string[];
}

export interface AttrRef {
    nodeId: string;
    endpointId: number;
    cluster: string;
    attribute: string;
}

export interface WriteAttrParams extends AttrRef {
    value: unknown;
}

export interface InvokeParams {
    nodeId: string;
    endpointId: number;
    cluster: string;
    command: string;
    args?: Record<string, unknown>;
}

export interface RemoveNodeParams {
    nodeId: string;
}

/**
 * How the node actually left.
 *
 * `decommissioned` — the device dropped our fabric and no longer lists us among
 * its connected services. `forced` — the device was unreachable, so it was only
 * forgotten locally and still carries our fabric; the user has to factory-reset
 * it to clean that up.
 */
export interface RemoveNodeResult {
    removed: "decommissioned" | "forced";
    message?: string;
}

// --- event payloads ---

export interface AttributeChanged {
    nodeId: string;
    endpointId: number;
    cluster: string;
    attribute: string;
    value: unknown;
}

export interface NodeLifecycle {
    nodeId: string;
}

/**
 * A Matter event (as opposed to an attribute change): a button press, an alarm
 * firing, a wash cycle finishing. Carries the cluster's own payload untouched.
 */
export interface DeviceEvent {
    nodeId: string;
    endpointId: number;
    cluster: string;
    event: string;
    data: unknown;
}

export interface SubscribeParams {
    /**
     * Highest seq the client has already processed. Omit (or 0) on a first
     * connect to ask for a full snapshot instead of a replay.
     */
    sinceSeq?: number;
}

export interface SubscribeResult {
    /**
     * The server's sequence position *before* this call did anything.
     *
     * A client must clamp its own counter to this value: if it is lower than
     * what the client remembers, the sidecar process restarted and its
     * numbering began again. Without the clamp the client would mistake every
     * new event for a duplicate and go silent.
     */
    seq: number;
    /** How many buffered events were resent on this socket. */
    replayed: number;
    /**
     * True when the requested position had already fallen out of the ring
     * buffer (or no position was given). A full snapshot was published
     * instead — the events follow this response's own frame order.
     */
    gap: boolean;
}

/**
 * Signalling relayed from a camera. Matter carries only this handshake — the
 * video itself flows over a WebRTC peer connection between the camera and
 * whoever holds the browser session, never through this sidecar.
 */
export type WebRtcSignal =
    | { kind: "offer"; sessionId: number; sdp: string }
    | { kind: "answer"; sessionId: number; sdp: string }
    | { kind: "ice"; sessionId: number; candidates: string[] }
    | { kind: "end"; sessionId: number; reason?: string };

export interface WebrtcOfferParams {
    nodeId: string;
    endpointId: number;
    /** SDP offer produced by the viewer's RTCPeerConnection. */
    sdp: string;
    /** Existing stream ids to reuse; omit to let the camera allocate. */
    videoStreamId?: number;
    audioStreamId?: number;
}

export interface WebrtcOfferResult {
    sessionId: number;
    videoStreamId?: number;
    audioStreamId?: number;
}

export interface WebrtcIceParams {
    sessionId: number;
    candidates: string[];
}

export interface WebrtcStopParams {
    sessionId: number;
}

export interface CommissioningProgress {
    stage: string;
    message?: string;
}

// --- errors ---
export const RpcErrorCode = {
    BadRequest: -32602,
    MethodNotFound: -32601,
    Internal: -32603,
    NotFound: -32001,
    NotReady: -32002,
    Unreachable: -32003,
    Timeout: -32004,
    Unsupported: -32005,
} as const;

/**
 * Error category. Keystone branches on this instead of matching message text,
 * so the wording of an error stays a presentation detail.
 */
export type ErrorKind =
    | "bad_request"
    | "not_found"
    | "not_ready"
    | "unreachable"
    | "timeout"
    | "unsupported"
    | "internal";

export interface RpcErrorBody {
    code: number;
    message: string;
    kind: ErrorKind;
    /** True when repeating the identical call could plausibly succeed. */
    retryable: boolean;
}

const ERROR_SHAPE: Record<ErrorKind, { code: number; retryable: boolean }> = {
    bad_request: { code: RpcErrorCode.BadRequest, retryable: false },
    not_found: { code: RpcErrorCode.NotFound, retryable: false },
    not_ready: { code: RpcErrorCode.NotReady, retryable: true },
    unreachable: { code: RpcErrorCode.Unreachable, retryable: true },
    timeout: { code: RpcErrorCode.Timeout, retryable: true },
    unsupported: { code: RpcErrorCode.Unsupported, retryable: false },
    internal: { code: RpcErrorCode.Internal, retryable: false },
};

/**
 * An error carrying its taxonomy. Throw this from controller methods; the WS
 * dispatcher turns it into the wire error body verbatim.
 */
export class RpcError extends Error {
    readonly kind: ErrorKind;
    readonly code: number;
    readonly retryable: boolean;

    constructor(kind: ErrorKind, message: string, opts?: { retryable?: boolean; cause?: unknown }) {
        super(message, opts?.cause !== undefined ? { cause: opts.cause } : undefined);
        this.name = "RpcError";
        this.kind = kind;
        this.code = ERROR_SHAPE[kind].code;
        this.retryable = opts?.retryable ?? ERROR_SHAPE[kind].retryable;
    }

    body(): RpcErrorBody {
        return { code: this.code, message: this.message, kind: this.kind, retryable: this.retryable };
    }
}

/**
 * Classify an arbitrary thrown value. Anything that isn't already an RpcError
 * came from matter.js or the runtime, so we inspect it for the failure modes
 * worth retrying — an unreachable or sleeping device is by far the most common,
 * and reporting it as "internal" would make keystone give up on a device that
 * is merely asleep.
 */
export function toRpcErrorBody(err: unknown): RpcErrorBody {
    if (err instanceof RpcError) return err.body();

    const message = err instanceof Error ? err.message : String(err);
    const name = err instanceof Error ? err.name : "";
    const haystack = `${name} ${message}`.toLowerCase();

    let kind: ErrorKind = "internal";
    if (haystack.includes("timeout") || haystack.includes("timed out")) {
        kind = "timeout";
    } else if (
        haystack.includes("unreachable") ||
        haystack.includes("no route") ||
        haystack.includes("channel") ||
        haystack.includes("disconnect") ||
        haystack.includes("session")
    ) {
        kind = "unreachable";
    }

    const shape = ERROR_SHAPE[kind];
    return { code: shape.code, message, kind, retryable: shape.retryable };
}

File: internal/adapters/matter/protocol.go
package matter

import (
	"encoding/json"
	"errors"
	"fmt"
)

// The matter.js sidecar speaks a small JSON-RPC dialect on the WebSocket:
//
//   client -> server:  {"id":"...", "method":"...", "params":{...}}
//   server -> client:  {"id":"...", "result":{...}}
//                 or:  {"id":"...", "error":{"code":..., "message":"..."}}
//                 or:  {"event":"...", "data":{...}}   (server-pushed)
//
// This file declares the wire types. Both ends must stay compatible.

// Method names. Kept as string constants so callers get compiler help.
const (
	MethodCommission     = "commission"
	MethodListNodes      = "listNodes"
	MethodReadAttribute  = "readAttribute"
	MethodWriteAttribute = "writeAttribute"
	MethodInvokeCommand  = "invokeCommand"
	MethodRemoveNode     = "removeNode"
	MethodSubscribe      = "subscribe"

	// MethodDiscoverCommissionable scans for devices advertising themselves as
	// ready to pair. Blocks for the requested window.
	MethodDiscoverCommissionable = "discoverCommissionable"
)

// Server-pushed event names.
const (
	EventAttributeChanged    = "attributeChanged"
	EventNodeOnline          = "nodeOnline"
	EventNodeOffline         = "nodeOffline"
	EventCommissioningStage  = "commissioningProgress"
	EventCommissionableFound = "commissionableFound"
	EventDeviceEvent         = "deviceEvent"
)

// Request is the client-to-server frame.
type Request struct {
	ID     string          `json:"id"`
	Method string          `json:"method"`
	Params json.RawMessage `json:"params,omitempty"`
}

// Response is the server-to-client reply frame. Exactly one of Result / Error
// is populated. Event is set only for server-pushed messages (no id, no
// method, no result — Event and Data instead).
type Response struct {
	ID     string          `json:"id,omitempty"`
	Result json.RawMessage `json:"result,omitempty"`
	Error  *RPCError       `json:"error,omitempty"`

	Event string          `json:"event,omitempty"`
	Data  json.RawMessage `json:"data,omitempty"`
	// Seq numbers server-pushed events within one sidecar process, starting
	// at 1. Zero means the sidecar predates sequencing.
	Seq int64 `json:"seq,omitempty"`
}

// SubscribeParams asks the sidecar to resync this connection.
type SubscribeParams struct {
	// SinceSeq is the highest seq we have already processed. Zero asks for a
	// full snapshot rather than a replay.
	SinceSeq int64 `json:"sinceSeq,omitempty"`
}

// SubscribeResult reports how the sidecar resynced us.
type SubscribeResult struct {
	// Seq is the sidecar's position before it handled our subscribe. We clamp
	// our own counter to it: a value below what we remember means the sidecar
	// process restarted and began numbering again, and without the clamp every
	// subsequent event would look like a duplicate and be dropped.
	Seq int64 `json:"seq"`
	// Replayed counts events resent to us from the sidecar's ring buffer.
	Replayed int `json:"replayed"`
	// Gap is true when the missed range was no longer buffered (or we had no
	// position) and the sidecar published a full snapshot instead.
	Gap bool `json:"gap"`
}

// ErrorKind classifies an RPC failure so callers can branch on the category
// instead of pattern-matching the message text.
type ErrorKind string

const (
	ErrKindBadRequest  ErrorKind = "bad_request" // caller sent something invalid; retrying won't help
	ErrKindNotFound    ErrorKind = "not_found"   // node/endpoint/attribute isn't there
	ErrKindNotReady    ErrorKind = "not_ready"   // controller still starting up
	ErrKindUnreachable ErrorKind = "unreachable" // peer didn't answer (asleep, off-network)
	ErrKindTimeout     ErrorKind = "timeout"     // operation exceeded its deadline
	ErrKindUnsupported ErrorKind = "unsupported" // device lacks this cluster/command
	ErrKindInternal    ErrorKind = "internal"    // sidecar bug or matter.js failure
)

// Sentinel errors for errors.Is. RPCError reports itself as matching the
// sentinel for its kind, so callers write:
//
//	if errors.Is(err, matter.ErrNotFound) { ... }
//
// instead of scanning the message for "not found".
var (
	ErrBadRequest  = errors.New("matter: bad request")
	ErrNotFound    = errors.New("matter: not found")
	ErrNotReady    = errors.New("matter: sidecar not ready")
	ErrUnreachable = errors.New("matter: device unreachable")
	ErrTimeout     = errors.New("matter: timeout")
	ErrUnsupported = errors.New("matter: unsupported operation")
	ErrInternal    = errors.New("matter: internal sidecar error")
)

var sentinelByKind = map[ErrorKind]error{
	ErrKindBadRequest:  ErrBadRequest,
	ErrKindNotFound:    ErrNotFound,
	ErrKindNotReady:    ErrNotReady,
	ErrKindUnreachable: ErrUnreachable,
	ErrKindTimeout:     ErrTimeout,
	ErrKindUnsupported: ErrUnsupported,
	ErrKindInternal:    ErrInternal,
}

// kindByCode derives a kind from the numeric code, for sidecars older than the
// kind field (and for the JSON-RPC codes that have a fixed meaning).
var kindByCode = map[int]ErrorKind{
	RPCCodeBadRequest:    ErrKindBadRequest,
	RPCCodeMethodMissing: ErrKindUnsupported,
	RPCCodeInternal:      ErrKindInternal,
	RPCCodeNotFound:      ErrKindNotFound,
	RPCCodeNotReady:      ErrKindNotReady,
	RPCCodeUnreachable:   ErrKindUnreachable,
	RPCCodeTimeout:       ErrKindTimeout,
}

// Numeric codes on the wire. The -32xxx range below -32000 is reserved by
// JSON-RPC for implementation-defined server errors.
const (
	RPCCodeBadRequest    = -32602
	RPCCodeMethodMissing = -32601
	RPCCodeInternal      = -32603
	RPCCodeNotFound      = -32001
	RPCCodeNotReady      = -32002
	RPCCodeUnreachable   = -32003
	RPCCodeTimeout       = -32004
	RPCCodeUnsupported   = -32005
)

// RPCError is the standard error object.
type RPCError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
	// Kind and Retryable are set by sidecars that speak the taxonomy. Older
	// ones omit them, so read via Kind() / IsRetryable() rather than directly.
	ErrKind   ErrorKind `json:"kind,omitempty"`
	Retryable bool      `json:"retryable,omitempty"`
}

func (e *RPCError) Error() string {
	if e == nil {
		return ""
	}
	if kind := e.Kind(); kind != "" {
		return fmt.Sprintf("%s (%s)", e.Message, kind)
	}
	return e.Message
}

// Kind returns the declared kind, falling back to one derived from the numeric
// code so a sidecar that predates the field still classifies correctly.
func (e *RPCError) Kind() ErrorKind {
	if e == nil {
		return ""
	}
	if e.ErrKind != "" {
		return e.ErrKind
	}
	return kindByCode[e.Code]
}

// IsRetryable reports whether repeating the call could plausibly succeed. The
// sidecar's explicit flag wins; otherwise the kind decides — a sleeping device
// or a controller that is still booting will answer later, a malformed request
// never will.
func (e *RPCError) IsRetryable() bool {
	if e == nil {
		return false
	}
	if e.Retryable {
		return true
	}
	switch e.Kind() {
	case ErrKindNotReady, ErrKindUnreachable, ErrKindTimeout:
		return true
	default:
		return false
	}
}

// Is makes errors.Is(err, ErrNotFound) work through wrapping.
func (e *RPCError) Is(target error) bool {
	if e == nil {
		return false
	}
	return sentinelByKind[e.Kind()] == target
}

// IsRetryable reports whether an error anywhere in the chain came back from the
// sidecar marked as worth retrying. Transport-level failures count too: a call
// that never reached a connected sidecar can be retried once it is back.
func IsRetryable(err error) bool {
	var rpcErr *RPCError
	if errors.As(err, &rpcErr) {
		return rpcErr.IsRetryable()
	}
	return errors.Is(err, ErrSidecarUnavailable) || errors.Is(err, ErrNotConnected)
}

// KindOf extracts the taxonomy kind from anywhere in an error chain, or "" if
// the error didn't originate from the sidecar.
func KindOf(err error) ErrorKind {
	var rpcErr *RPCError
	if errors.As(err, &rpcErr) {
		return rpcErr.Kind()
	}
	return ""
}

// --- Method params / results ---

// CommissionParams carries the setup code the user read off the device (manual
// 11/21-digit code or an "MT:" QR payload) and, optionally, which discovered
// device to pair.
type CommissionParams struct {
	SetupCode string `json:"setupCode"`
	// Target is a ref from a previous discoverCommissionable scan. Empty means
	// "find the device from the setup code", which is what a user typing a code
	// with no prior scan gets.
	Target string `json:"target,omitempty"`
}

// DiscoverCommissionableParams bounds a scan.
type DiscoverCommissionableParams struct {
	TimeoutMs int64 `json:"timeoutMs,omitempty"`
}

// CommissionableDevice is one advertisement reported by a scan. Also the
// payload of the commissionableFound event, pushed while the scan is running.
type CommissionableDevice struct {
	Ref           string `json:"ref"`
	Name          string `json:"name,omitempty"`
	DeviceType    string `json:"deviceType,omitempty"`
	VendorID      int    `json:"vendorId,omitempty"`
	ProductID     int    `json:"productId,omitempty"`
	Discriminator int    `json:"discriminator,omitempty"`
}

// CommissionResult is what the sidecar reports after joining the fabric.
type CommissionResult struct {
	NodeID      string `json:"nodeId"`
	FabricIndex int    `json:"fabricIndex"`
}

// Node is a single Matter node reported by listNodes. Endpoints and clusters
// are what the engine maps into keystone Features via mapping.go.
type Node struct {
	NodeID      string `json:"nodeId"`
	VendorName  string `json:"vendorName,omitempty"`
	ProductName string `json:"productName,omitempty"`
	VendorID    int    `json:"vendorId,omitempty"`
	ProductID   int    `json:"productId,omitempty"`
	// NodeLabel is BasicInformation.NodeLabel — the name the user gave the
	// device in whichever ecosystem commissioned it first (Apple Home, Google
	// Home). Usually friendlier than VendorName + ProductName.
	NodeLabel string     `json:"nodeLabel,omitempty"`
	Online    bool       `json:"online"`
	Endpoints []Endpoint `json:"endpoints"`
}

// Endpoint describes one addressable endpoint on a node. DeviceType is the
// canonical CSA device-type name the sidecar resolved from
// Descriptor.DeviceTypeList ("OnOffLight", "ColorTemperatureLight", …) — empty
// if the endpoint hasn't been interviewed or declares a type we don't know.
type Endpoint struct {
	EndpointID int      `json:"endpointId"`
	DeviceType string   `json:"deviceType,omitempty"`
	Clusters   []string `json:"clusters"`
}

// AttrRef locates a single attribute for read/write operations.
type AttrRef struct {
	NodeID     string `json:"nodeId"`
	EndpointID int    `json:"endpointId"`
	Cluster    string `json:"cluster"`
	Attribute  string `json:"attribute"`
}

// WriteAttrParams is AttrRef plus the desired value.
type WriteAttrParams struct {
	AttrRef
	Value any `json:"value"`
}

// InvokeParams targets a cluster command on a specific endpoint.
type InvokeParams struct {
	NodeID     string         `json:"nodeId"`
	EndpointID int            `json:"endpointId"`
	Cluster    string         `json:"cluster"`
	Command    string         `json:"command"`
	Args       map[string]any `json:"args,omitempty"`
}

// RemoveNodeParams identifies a node to unpair from the local fabric.
type RemoveNodeParams struct {
	NodeID string `json:"nodeId"`
}

// --- Event payloads ---

// AttributeChanged is the payload of the attributeChanged server event.
type AttributeChanged struct {
	NodeID     string `json:"nodeId"`
	EndpointID int    `json:"endpointId"`
	Cluster    string `json:"cluster"`
	Attribute  string `json:"attribute"`
	Value      any    `json:"value"`
}

// DeviceEvent is the payload of the deviceEvent server event: a Matter event
// rather than an attribute change (button press, alarm, cycle completion).
type DeviceEvent struct {
	NodeID     string `json:"nodeId"`
	EndpointID int    `json:"endpointId"`
	Cluster    string `json:"cluster"`
	Event      string `json:"event"`
	Data       any    `json:"data,omitempty"`
}

// NodeLifecycle is the payload of nodeOnline / nodeOffline.
type NodeLifecycle struct {
	NodeID string `json:"nodeId"`
}

// CommissioningProgress is emitted while a commission() call runs.
type CommissioningProgress struct {
	Stage   string `json:"stage"`
	Message string `json:"message,omitempty"`
}
